docs/src/lib/functions.md

```diff
@@ -48,4 +48,62 @@ sort!
 tail
 unique!
 permutecols!
-```
\ No newline at end of file
+```
+
+## Indexing
+
+The following rules explain target functionality of how `getindex` and `view` are intended to work for `DataFrame`, `SubDataFrame` and `DataFrameRow`.
+
+The rules for a valid type of index into a column are the following:
+* a value, later denoted as `col`:
+    * a `Symbol`;
+    * an `Integer` that is not `Bool`;
+* a vector, later denoted as `cols`:
+    * a vector of `Symbol` (does not have to be a subtype of `AbstractVector{Symbol}`);
+    * a vector of `Integer` other than `Bool` (does not have to be a subtype of `AbstractVector{<:Integer}`);
+    * a vector of `Bool` that has to be a subtype of `AbstractVector{Bool}`.
+    * a colon.
+
+The rules for a valid type of index into a row are the following:
+* a value, later denoted as `row`:
+    * an `Integer` that is not `Bool`;
+* a value, later denoted as `rows`:
+    * a vector of `Integer` other than `Bool` (does not have to be a subtype of `AbstractVector{<:Integer}`);
+    * a vector of `Bool` that has to be a subtype of `AbstractVector{Bool}`;
+    * a colon.
+
+The following list specifies return types of `getindex` operations depending on argument types. In the summary `df` represents a `DataFrame`, `sdf` is a `SubDataFrame` and `dfr` is a `DataFrameRow`.
+
+`DataFrame`:
+* `df[col]` -> an alias of a vector contained in column `col`;
+* `df[cols]` -> a freshly allocated `DataFrame` containing aliases of vectors contained in columns `cols`;
+* `df[row, col]` -> a value contained in row `row` of column `col`;
+* `df[row, cols]` -> a `NamedTuple` containing data from row `row` in columns `cols`;
+* `df[rows, col]` -> a copy of a vector `df[col]` with only rows `rows` selected;
+* `df[rows, cols]` -> a `DataFrame` containing columns `cols` and `df[rows, col]` as a vector in each `col` in `cols`.
+* `@view df[col]` -> an alias of a vector contained in column `col`;
+* `@view df[cols]` -> a `SubDataFrame` with parent `df` if `cols` is a colon and `df[cols]` otherwise;
+* `@view df[row, col]` -> translates to `view(df[col], row)` (a `0`-dimensional view into `df[col]`);
+* `@view df[row, cols]` -> a `DataFrameRow` with parent `df` if `cols` is a colon and `df[cols]` otherwise;
```

Comment:
So this is going to require creating a new `DataFrame` for each row? That sounds really bad. Maybe we could change `DataFrameRow` to also take a subset of rows? That's really a corner case though, so it could be fixed later.

Reply:
We have the same situation with `SubDataFrame` as current `SubDataFrame` and `DaraFrameRow` were designed with the assumption that only row-indexing is performed.

I would leave it for a separate PR (we will have to handle colon and subset of columns separately and support a mapping from column numbers of a view to column numbers of a parent in the latter). In the worst case user should first create a parent that has correct columns and then run a `view` on all of them. I will open an issue for this to keep track of it.

Reply:
`https://github.com/JuliaData/DataFrames.jl/issues/1557`